Ethernet support for STM32F429

Ioan-Cristian opened the pull request:
### Pull Request Overview

This pull request adds basic Ethernet support for STM32F429. At the moment, the
peripheral supports only raw Ethernet frames and there is no support for
address filtering, hardware checksum control or chaining descriptors for Ethernet DMA.

### Testing Strategy

The Ethernet implementation was tested through basic unit tests and manually running a web server. (lwIP)

### TODO or Help Wanted

Issues:

+ In rare situations, the peripheral (or the capsule?) hangs and no frame is
processed anymore.
+ When compiling with level 2 or 3 optimizations, the web server doesn't work.

TODO:

+ Try to successfully compile and run the web server with optimizations (high priority)
+ Full UDP stack in the kernel
+ Offload workload from software to hardware
+ Chaining descriptors for Ethernet DMA.
+ Address filtering
+ Packets statistics: drop frames, bad frames, received/transmitted frames etc
for profiling

### Documentation Updated

- [+] Updated the relevant files in `/docs`

### Formatting

- [+] Ran `make prepush`.

### Updates

+ 29/06/2023: This pull request was converted to a draft, since the clock implementation was split up in #3528

Ioan-Cristian commented on boards/weact_f401ccu6/layout.ld:
```diff
@@ -1,13 +1,13 @@
- * user = 128KB
- * ram = 64KB */
+ * rom = 256KiB (LENGTH = 0x00040000)
+ * kernel = 150KiB
+ * user = 102KiB
+ * ram = 64KiB */
 
 MEMORY
 {
-  rom (rx)  : ORIGIN = 0x08000000, LENGTH = 0x00020000
-  prog (rx) : ORIGIN = 0x08020000, LENGTH = 0x00020000
+  rom (rx)  : ORIGIN = 0x08000000, LENGTH = 150K
+  prog (rx) : ORIGIN = 0x08025800, LENGTH = 102K
   ram (rwx) : ORIGIN = 0x20000000, LENGTH = 64K
 }
 
```
The total flash size for the kernel was increased because otherwise the kernel couldn't be compiled in debug mode. This problem seems to also occur on the `master` branch.

Ioan-Cristian commented on capsules/extra/src/ethernet_app_tap.rs:
```diff
@@ -213,7 +213,7 @@ pub const DRIVER_NUM: usize = capsules_core::driver::NUM::EthernetTAP as usize;
 /// Maximum size of a packet which can be transmitted over the underlying
 /// [`EthernetAdapter`] device. Currently hard-coded to `1522` bytes.
-pub const MAX_MTU: usize = 1522;
+pub const MAX_MTU: usize = 1524;
```
Receive descriptors allow buffer lengts multiple of 4, 8 or 16, depending how the DMA is configured.

lschuermann replied:
This is a good point. We should reconsider this in the design when we revisit the packet buffer passing story around Tock.

brghena replied:
The comment here needs to be updated. Probably worth noting the length limitations of 4/8/16

Ioan-Cristian replied:
Updated comment. Probably this constant will be removed from the capsule and instead queried from the peripheral itself once the Networking Working Group establish a HIL for that.

Ioan-Cristian commented on chips/stm32f4xx/Cargo.toml:
```diff
@@ -12,3 +12,9 @@ edition.workspace = true
 cortexm4 = { path = "../../arch/cortex-m4" }
 enum_primitive = { path = "../../libraries/enum_primitive" }
 kernel = { path = "../../kernel" }
+
+[features]
+stm32f401 = []
+stm32f412 = []
+stm32f429 = []
+stm32f446 = []
```
To avoid a lot of code duplication, conditional compilation was preferred.

Ioan-Cristian commented on chips/stm32f4xx/src/flash.rs:
```diff
@@ -0,0 +1,711 @@
+use kernel::utilities::StaticRef;
+use kernel::ErrorCode;
+
+register_structs! {
```
This struct represents the common superset of all chips Flash. The field that is not available on some chips shouldn't be accessed. This was done to avoid code duplication.

Ioan-Cristian commented on chips/stm32f4xx/src/flash.rs:
```diff
@@ -0,0 +1,711 @@
+    }
+}
+
+register_bitfields![u32,
```
This is the common superset of all possible bitfieds for all chips. This was done to avoid code duplication.

bradjc reviewed (changes requested):
Blocking because of the added nightly feature which since we want to build Tock on stable rust we would be hesitant to add.

Generally this looks good, and it's exciting to see this implementation!

I have to say, the cargo features make me uneasy. I'm not entirely against adding them because _when_ to use them is not ambiguous: either the kernel is being compiled for a specific chip or its not. They aren't features in the sense that different developers might turn them on or off to get different functionality.

However, they do present an opportunity for confusion: is the MCU variant code in the variant crate, or in the general chip crate guarded by a feature? In general, to avoid this confusion we put shared code in the general chip crate, and then "customize it" (often for example by setting the MMIO address of a peripheral) in the variant-specific crate. I think that would be our preference.

However, there may still be reasons to use features. Another way to address the confusion about where all the variant-specific customization is is to group the settings in one (and only one) place, like we did here: https://github.com/tock/tock/blob/master/chips/earlgrey/src/chip_config.rs. That way, the feature is only "used" once, it is clear where to document all of the differences and why the differences need to be included in Tock, and the remaining code reads as traditional normal rust code.

What I would like to avoid is having features used in many places haphazardly.

Also, I'm not worried about features in the tests.

lschuermann commented on boards/nucleo_f429zi/src/main.rs:
```diff
@@ -44,7 +50,7 @@ const FAULT_RESPONSE: kernel::process::PanicFaultPolicy = kernel::process::Panic
 #[no_mangle]
 #[link_section = ".stack_buffer"]
-pub static mut STACK_MEMORY: [u8; 0x2000] = [0; 0x2000];
+pub static mut STACK_MEMORY: [u8; 0x4000] = [0; 0x4000];
```
This is expected because of my terribly inefficient `tap_driver` implementation on the `tock-ethernet` branch. This PR is targeting this branch, and so we should focus on the diff concerning the added driver here. When it comes time to merge `tock-ethernet` into master, I think it's a good opportunity to consider the remaining infrastructure on that branch. Things are still WIP there.

lschuermann commented on chips/stm32f429zi/src/ethernet/mod.rs:
```diff
@@ -0,0 +1,2328 @@
+
+register_structs! {
+    /// Ethernet: media access control
+/// (MAC)
```
We should fix the formatting in those register definitions.

Ioan-Cristian replied:
Fixed formatting.

lschuermann commented on chips/stm32f429zi/src/ethernet/mod.rs:
```diff
@@ -0,0 +1,2328 @@
+}
+
+register_bitfields![u32,
+MACCR [
+    /// RE
+    RE OFFSET(2) NUMBITS(1) [],
```
Perhaps fix the indentation here as well. It's unfortunate that `rustfmt` doesn't really format our custom macros well.

lschuermann commented on chips/stm32f429zi/src/ethernet/mod.rs:
```diff
@@ -0,0 +1,2328 @@
+        debug!("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
+        debug!("Testing Ethernet initialization...");
+
+        // This is broken for some reasons
+        assert_eq!(Ok(()), ethernet.init());
```
If this is broken, we probably shouldn't ship this test? Do you know what the underlying root cause is?

lschuermann reviewed (changes requested):
I did a preliminary review of these changes, the Ethernet driver code looks pretty good! Certainly an impressive feat to get this complex of a MAC under control. :+1: This is still lacking a pass over the code to address @bradjc's comments, and the TODOs and formatting through the changeset.

To give some more context for these changes: this is part of #3308. The PR is made against the `tock-ethernet` branch, which already contains some initial in-kernel infrastructure and a minimal "Ethernet HIL" to support userspace network stacks (e.g., currently an LwIP-based webserver in `libtock-c`, `smoltcp` integration in `libtock-rs` is WIP).

I think it would help for the STM clocking support to be split out of this changeset (perhaps using git filter) and PRed against the upstream master branch directly. I can then rebase `tock-ethernet` to incorporate those changes and reduce the diff in this PR.

Ioan-Cristian commented:
I am going to close this pull request and create two new ones: one for the clocks and one for the Ethernet as suggested. Both of them should be ready in the coming hours.

Ioan-Cristian commented:
The clock implementation was split up in a new pull request. For all clock related issues, see #3528 .

Ioan-Cristian commented on chips/stm32f4xx/src/lib.rs:
```diff
@@ -9,6 +9,7 @@
 #![crate_name = "stm32f4xx"]
 #![crate_type = "rlib"]
 #![no_std]
+#![feature(core_panic)]
```
No longer related to this pull request.

Ioan-Cristian commented on boards/weact_f401ccu6/layout.ld:
```diff
@@ -1,13 +1,13 @@
-  rom (rx)  : ORIGIN = 0x08000000, LENGTH = 0x00020000
-  prog (rx) : ORIGIN = 0x08020000, LENGTH = 0x00020000
+  rom (rx)  : ORIGIN = 0x08000000, LENGTH = 150K
+  prog (rx) : ORIGIN = 0x08025800, LENGTH = 102K
```
No longer related to this pull request.

Ioan-Cristian commented on boards/nucleo_f429zi/src/main.rs:
```diff
@@ -16,13 +16,19 @@ use capsules_core::virtualizers::virtual_alarm::VirtualMuxAlarm;
 use stm32f429zi::interrupt_service::Stm32f429ziDefaultPeripherals;
+#[allow(unused_imports)]
+use stm32f429zi::rcc::{APBPrescaler, MCO1Source, SysClockSource};
+use stm32f429zi::syscfg::EthernetInterface;
```
Removed `#[allow(unused_imports)]`.

Ioan-Cristian commented on boards/nucleo_f429zi/src/main.rs:
```diff
@@ -288,6 +299,63 @@ unsafe fn setup_peripherals(
 }
 
+// Helper function to setup Ethernet pins
+#[allow(dead_code)]
```
Removed `#[allow(dead_code)]`.

alevy commented:
@Ioan-Cristian I think I shot you in the foot a bit by pushing directly to the `tock-ethernet` branch a merge with master, so _this_ PR now has >400 files changed compared to current `tock-ethernet` and >700 commits. This was my mistake, but I think we can fix this with another merge or cherry picking your commits here onto `tock-ethernet`

lschuermann commented:
Sorry for the confusion here! We merged `master` into `tock-ethernet` to keep it up to date, which is reasonable for a shared development branch. However, this has seemed to result in this PR including all of these commits in its diff as well.

I have verified locally that git is clever enough to produce a merge commit which only includes the files actually changed in this PR. Instead it appears that this PR's diff can be explained by [GitHub's three-dot diff](https://stackoverflow.com/a/74871448) algorithm: whereas a `git diff $TARGET_BRANCH..$PR_BRANCH` would show you the full delta between the merge target and the PR (and thus also include changes made as the target moves forward), GitHub picks a common ancestor on both the target branch and PR branch. It then effectively displays a `git diff $COMMON_ANCESTOR..$PR_BRANCH`. However, this is not what a merge-commit would contain: merges can detect duplicate commits and identical changes, and so the actual merge-diff for this PR indeed solely contains the actual changes.

We should be able to fix this PR's diff by either rebasing this branch on `tock-ethernet`, or merging `tock-ethernet` into this branch. The latter seems like a more reasonable option, given this branch also contains merge-commits.

lschuermann commented on chips/stm32f429zi/src/ethernet/ethernet.rs:
```diff
@@ -0,0 +1,2288 @@
+
+register_structs! {
+    /// Ethernet: media access control (MAC)
+    Ethernet_MacRegisters {
```
Nit: usually, names in Rust follow either "snake_case", "UPPER_CASE_SNAKE_CASE" or "PascalCase", but not a mixture of those. Typically we try to stick to those, but here it might not matter: the tock-registers macros generate the appropriate `#[allow(_)]` markers for names to violate these rules (as we use the names for both modules and structs).

Ioan-Cristian replied:
Fixed names.

lschuermann commented on chips/stm32f429zi/src/ethernet/ethernet.rs:
Note to self: This is safe, given that both `TransmitDescriptor` and `ReceiveDescriptor` are `#[repr(C)]` structs which only contain `InMemoryRegister`s, which in turn wrap `VolatileCell` -- so we don't need to main Rust's aliasing requirements here, and spurious reads are also not an issue.

Ioan-Cristian replied:
Added comment with your explanation.

lschuermann commented on chips/stm32f429zi/src/ethernet/ethernet.rs:
We should probably resolve these comments and add these functions or remove them appropriately. More generally, I think it's perfectly fine for a driver to not expose an exhaustive set of hardware capabilities, especially as accessor methods. Such methods typically suggest that some functionality is either required, or at least supported in a driver. Not implementing them can be interpreted as "we don't know whether touching this bit is safe, so we don't provide a method for it".

Ioan-Cristian replied:
> we don't know whether touching this bit is safe, so we don't provide a method for it

That's exactly the issue. The general description of DMA interrupts (page 1189):

> When all the enabled interrupts within a group are cleared, the corresponding summary bit is cleared.

The page describing registers (1226):

> The normal interrupt summary bit value is the logical OR of the following when the corresponding interrupt bits are enabled in the ETH_DMAIER register

So I far I understand that the summary bit is automatically cleared. However, the same page describing summary bits:

> This is a sticky bit and it must be cleared (by writing a 1 to this bit) each time a corresponding bit that causes NIS to be set is cleared
> This is a sticky bit and it must be cleared each time a corresponding bit that causes AIS to be
set is cleared

Do I miss something?

lschuermann commented on chips/stm32f429zi/src/ethernet/ethernet.rs:
Same here, resolve TODO.

Ioan-Cristian replied:
See above.

lschuermann commented on chips/stm32f429zi/src/ethernet/ethernet.rs:
In general, Tock drivers typically use the tock-registers interface directly to query such bits -- accessor methods like these are mostly used in case of some more complex bit-queries / bit-manipulations. However, in the form of private methods they're also totally fine to have around; feel free to leave them in if that's preferred.

Ioan-Cristian replied:
I find it's easier to understand what code does when using methods with explicit names compared to registers named `dmasr`, especially that the Tock guideline suggests keeping the name of the registers the same as in the reference manual. However, I'm not opposed to remove them if you wish so.

lschuermann replied:
It seems fine to keep them, just a style choice.

lschuermann commented on chips/stm32f429zi/src/ethernet/ethernet.rs:
For such methods featuring a timeout, it'd be great to document the exact reasons why this could fail, what the magic timeout parameter means (i.e. why you came up with this value, even if it just happens to work in practice that's worth documenting), and perhaps make it configurable via a parameter.

alevy replied:
Alternatively, if this is really guaranteed to succeed _eventually_ it's OK to loop until the condition is true. E.g., some of the GPIO hardware (such as the SAM4L) have a busy bit which will eventually clear, so the drivers for that hardware just loops on the busy bit before reading or setting the GPIO state and there is no error condition.

Ioan-Cristian replied:
Some of these for loops can fail if the peripheral mode configuration register doesn't match the actual hardware. Not sure if I should replace them with while loops since by the time the MAC and DMA cores are initialized, `syscfg` is already configured in the right mode. The sole benefit of keeping a for loop would be spotting easier the case where the syscfg is not configured correctly, since this DMA/MAC core seems to be present on multiple boards Tock supports (MSP432 for instance).

lschuermann commented on chips/stm32f429zi/src/ethernet/ethernet.rs:
I think it may be better to not keep such unused functions around, assuming you don't plan on integrating them in the near term. If required, we can always re-add them later.

lschuermann commented on chips/stm32f429zi/src/ethernet/ethernet.rs:
Same here, explain timeout.

lschuermann commented on chips/stm32f429zi/src/ethernet/ethernet.rs:
Perhaps we should avoid panicing here in the long term, except if these are errors which really put the chip in an unrecoverable state. However, something like a "transmit buffer underflow" seems like the kind of interrupt which could be occurring at runtime without major impact (except for a broken Ethernet frame), depending on the AHB bus.

Ioan-Cristian replied:
Removed most of the panics and added additionally comments to explain the intent of the panic handler actions. Might require future revision.

lschuermann commented on chips/stm32f429zi/src/ethernet/ethernet.rs:
I believe these functions are no longer relevant? In general, a receive API will likely be asynchronous, with the MAC maintaining a (shared) buffer poll which it can pass to a registered client on demand.

Ioan-Cristian replied:
The story is that when I started working on the Ethernet, I was not aware of the `tock-ethernet` branch, so I started to create my own Ethernet HIL. Since Ethernet is another mean for transmitting and receiving data, I mimicked the UART HIL. Now, all traits are removed from the HIL.

lschuermann reviewed (changes requested):
My apologies for the (significantly) delayed review! I did my first full thorough pass through the `stm32f429zi/src/ethernet/ethernet.rs`, and overall it looks great! I especially appreciate the clean separation of low-level methods, which essentially just map hardware-operations to Rust methods, and high-level methods that configure the hardware & implement TX / RX.

I left a few comments, many of which are nits. The driver contains a lot of simple accessor methods, many of which are only used once or twice; those could be inlined, but that's a question of style.

More generally, I don't think the additions to the preliminary WIP Ethernet HIL in this PR are objectionable, but they certainly warrant further discussion. I presume we'd workshop them as part of the Network WG, and under consideration of the other Ethernet MACs, but that is outside of the scope of this PR.

Assuming #3528 gets in (which I hope to be able to take a look at next), I think this is pretty solid in terms of its driver implementation. I don't see many blockers for the other (HIL) changes to land in `tock-ethernet` as well.

Ioan-Cristian commented:
Addressed most of the comments. However, I don't have any board near me to test the changes, so I suggest you to wait until tomorrow before reviewing them. I'll inform you when testing is done.

Ioan-Cristian commented:
Completely removed any addition from the Ethernet HIL.
lwIP and smoltcp apps still working.
The only `#[allow(dead_code)]` functions that I kept are those for disabling the Ethernet peripheral, but you think there will never be a need to disable the Ethernet peripheral once started, I can delete them.
I hope I didn't forget to address any comment. If I did, please tell me.
Waiting for your other comments/issues. Thank you!

Ioan-Cristian commented:
Hello! Sorry for my long inactivity! Is there a possibility to merge the upstream master branch?

Ioan-Cristian commented:
I managed to cherry pick all the relevant commits to reduce the number of changed files. Would it be possible to merge the `master` branch into `tock-ethernet` branch? The Ethernet implementation relies on clocks.

bradjc commented:
@lschuermann
